include/networkit/algebraic/CSRGeneralMatrix.hpp

```diff
@@ -0,0 +1,1251 @@
+/*
+ * CSRGeneralMatrix.hpp
+ *
+ *  Created on: May 6, 2015
+ *     Authors: Michael Wegner <michael.wegner@student.kit.edu>
+ *              Eugenio Angriman <angrimae@hu-berlin.de>
+ */
+
+// networkit-format
+
+#ifndef NETWORKIT_ALGEBRAIC_CSR_GENERAL_MATRIX_HPP_
+#define NETWORKIT_ALGEBRAIC_CSR_GENERAL_MATRIX_HPP_
+
+#include <algorithm>
+#include <numeric>
+#include <omp.h>
+#include <vector>
+
+#include <networkit/Globals.hpp>
+#include <networkit/algebraic/AlgebraicGlobals.hpp>
+#include <networkit/algebraic/SparseAccumulator.hpp>
+#include <networkit/algebraic/Vector.hpp>
+#include <networkit/auxiliary/Timer.hpp>
+#include <networkit/graph/Graph.hpp>
+
+namespace NetworKit {
+
+/**
+ * @ingroup algebraic
+ * The CSRGeneralMatrix class represents a sparse matrix stored in CSR-Format
+ * (i.e. compressed sparse row).
+ * If speed is important, use this CSRGeneralMatrix instead of the Matrix class.
+ */
+template <class ValueType>
+class CSRGeneralMatrix {
+    std::vector<index> rowIdx;
+    std::vector<index> columnIdx;
+    std::vector<ValueType> nonZeros;
+
+    count nRows;
+    count nCols;
+    bool isSorted;
+    ValueType zero;
+
+    /**
+     * Quicksort algorithm on columnIdx between [@a left, @a right].
+     * @param left
+     * @param right
+     */
+    void quicksort(index left, index right) {
+        if (left >= right)
+            return;
+        index pivotIdx = partition(left, right);
+        if (pivotIdx != 0) {
+            quicksort(left, pivotIdx - 1);
+        }
+        quicksort(pivotIdx + 1, right);
+    }
```

Comment:
We do we use an ad-hoc quicksort instead of using the standard library?

Reply:
A possible reason for this is because we need to sort two vectors at once (columnIdx and nonZero). We could replace this ad-hoc quicksort by computing the permutation of indices using `std::sort` and then apply the permutation to the two vectors as done in #734 (but to do so we might want to merge the `applyPermutation` function in #734 first).
Do you know if there is a better way to do this?